README.md
# GW2 DPS Log Uploader

A command line tool used to pick specific log and upload to [Gw2Raidar](https://www.gw2raidar.com) and [dps.report](https://dps.report/).

[中文說明](https://github.com/paddycup1/Gw2DpsLogUploader/blob/master/中文說明.md)

## Download

[Release Page](https://github.com/paddycup1/Gw2DpsLogUploader/releases)

## Initialize

Run `DpsLogUploader.exe -init username password`  
username and password is used to get account token from Gw2Raidar.  
This command will generate `Config.json` in current directory, please open it with any text editor, update your acrdps log path into `LogPath` field.  
You should seperate your path with double back slash (`\\`) because I use json format. ex: `C:\\Users\\username\\Documents\\Guild Wars 2\\addons\\arcdps\\arcdps.cbtlogs`  
The field `Gw2RaidarToken` should be automatically filled by `-init` command.  
The field `DiscordBotToken` is optional, this is used by `EmbedHelper.exe` to print embed report to discord channel. It can be any value or deleted if `EmbedHelper.exe` doesn't used.

## Usage

`DpsLogUploader.exe [arguments]`  

There are two things need to give with arguments: which bosses and which time section  
Select bosses with `-boss` command.  
Select time section with `-starttime` + `-endtime` combination or `-past`, `-today`, `-yesterday`, `-alltime`.

**known issue: can't determine Deimos's encounter result.**  

## Supported Arguments

All Arguments is case insensitive except the player name and embed title/description.  
Arguments order is meaningless, just put all argument you want in any order.  

### Select boss

* `-boss name1 name2 ...` (`-b`) : Select bosses names need to be uploaded. Use names or alias defined in `BossList.json`.Bosses, or use group name defined in `BossList.json`.Groups to select multiple boss.  

### Select time section
* `-starttime [yy/]mm/dd [h:m[:s]]` (`-start`) : Select start point of target time section.
* `-endtime [yy/]mm/dd [h:m[:s]]` (`-end`) : Select end point of target time section.
* `-past 5h` or `-past 8d` (`-p`) : Select target time section from now back to given duration.
* `-today` : Select target time section from current day 00:00:00 to 23:59:59.
* `-yesterday` : Select target time section from yesterday day 00:00:00 to 23:59:59.
* `-raidreset` : Select target time section to current RAID reset.
* `-dailyreset` : Select target time section to current Daily reset.
* `-week` : Select target time section from current week Monday 00:00:00 to next Sunday 23:59:59.
* `-alltime` : Ignore time limit.

### Miscellaneous
* `-longest` : Only upload the longest log in each boss after filtered by other arguments, can't be used with `-last`.
* `-longerthan 5m[6s]` : Select the logs that fight time longer than given time.
* `-shorterthan 5m[6s]` : Select the logs that fight time shorter than given time.
* `-with name1 name2...` : Select the logs that contain given player name(can be either character name or display name).
* `-win` and `-fail`: Select encounter result to be uploaded, the tool will upload both if not selected. 
* `-healthover 54.32`: Filter logs by the boss's health, 100.00 ~ 0.00.
  * Note that above command will let the tool entirely parse every log which in selected boss and time section, **this will significantly increase the time cost**.
* `-last` : Only upload the last log in each boss after filtered by other arguments, can't be used with `-longest`.
* `-sort type` : Sort output result by given type, type can be `name`, `time` or `encounter`.
* `-reverse` (`-r`) : Reverse the sort result if `-sort` arguments is given.
* `-gen [rh] [ei] [raidar]` : Select upload target, default is all select.
  * `rh` is Raid Heroes host by dps.report
  * `ei` is Elite Insight host by dps.report
  * `raidar` or can be `rd` for short, is Gw2Raidar
* `-o filename` : Output result to given file name. Default is `output/output.txt` if not given.
* `-raidarlogin username password` : Retrieve account token from Gw2Raidar API. The token is needed for upload to Gw2Raidar.
* `-raidarwaittime 20` : Waiting time between retrying to get Raidar's encounter list, in second. The default value is 20.  
* `-raidarretrycount 15` : Retry times when getting Raidar's encounter list. The default value is 15.  
* `-raidarsearchcount 100` : The amount of logs get from Raidar in each try, this value should bigger than the log amount you uploaded. The default value is 100.  
* `-notonlyraidar` or `-nord` : Upload the logs to dps.report even not in Raidar's support list. If this argument is not given, logs not in Raidar's support list will be skip.  
* `-init username password` : Generate default config file and retrieve Gw2Raidar account token by given username and password.  
* `-json` : Gererate json format output.
* `-embed title [description] [color code(#FFFFFF)]` : Generate discord embed format output, you can use `EmbedHelper.exe` in this repository to print embed to your discord server.  
![](https://i.imgur.com/8I4NB5D.png)

## Example

#### Upload today's fotm log
`DpsLogUploader.exe -b fotms -past 1d -win`

#### Upload all success raid log in August and sort result with fight time
`DpsLogUploader.exe -b raids -start 2018/8/1 -end 2018/8/31 -win -sort time`

#### Upload all raid wing 4 log to Gw2Raidar and Raid Heroes
`DpsLogUploader.exe -b W4 -alltime -gen rd rh`

## Explation of BossList.json
You can customer boss alias by edit this file:
```json
{
  "Bosses": [
    {
      "Name": "The exactly Boss name in arcdps log folder, the tool will use this name to open log. This name can be used in -boss argument",
      "Aliases": [
        "Any aliases you want, any name defined here can be used in -boss arguments"
      ],
      "LifeThreshold": "This value should be a number, please check below section before change this value."
    }
  ],
  "Groups": [
    {
      "Name": "Any name you want, this name is used to refer group in -boss arguments",
      "Bosses": [
        "Name field of boss you want to add into this group. Don't use alias here"
      ]
    }
  ]
}
```

## LifeThreshold in BossList.json

This value affects how the tool judging a log is success or fail.  
If this value is not given, the tool will judge the result by the boss dead event.  (This works in the most fo bosses, but has some exception, like Arkk,he actually doesn't **die** when you win the combat)
If this value is given, the tool will judge the result by health update event, when the boss's health drop below the given value, you have won the combat. The unit of the value is 0.01%, so 10000 means 100%, 50 means 0.5%.  
To know how to set this value, please use the tool attach in the package, `EvtcParser.exe`. Execute `EvtcParser.exe LOG_FILE` will generate a simple report names `EvtcLog.txt`, or `EvtcParser.exe LOG_FILE -json` will dump full log content to `EvtcLog.json`. (It may be a little big)  
This tool will shows `Last Life Change` and `Is Boss Dead` at the top of the file, so you can set LifeThreshold according to this.  

* If `Is Boss Dead` is true, just remove this field or set it to a negative value.
* If `Is Boss Dead` is false, please check `Last Life Change`, this value means the last life change event record in log, please set `LifeThreshold` to a value higher but close to it. (I'll set about 20~50 higher than `Last Life Change`)

Or execute `EvtcParser.exe LOG_FILE -config`, and give it **a success log** then it will config it automatically.

# Embed Helper

A very very simple discord bot only used to print custome embed structure to discord.

## Initialize

Open `Config.json` generate by `DpsLogUploader.exe -init username password` or make it by your self. The only one requirement is `DiscordBotToken` field in this file.  
If you haven't  created a discord bot ↓↓↓  
1. Login [discord developer page](https://discordapp.com/developers/applications/)
2. Create your own application.
3. Create a bot in application manage page.
4. Copy bot token in bot tab into `DiscordBotToken` field in `Config.json`.
5. Invite your bot in your server by OAuth2 URL generated by OAuth2 tab in application manage page.

## Usage
Activate bot by command `EmbedHelper.exe`  
In discord channel, send `!embed ` followed by entire json embed code.  
![](https://i.imgur.com/sBiGXOl.png)  
The bot will delete this message immediately then send a embed message.  
![](https://i.imgur.com/ttH0O3E.png)  

If your embed code is too big to send to discord, you can send file name with command:  
`!embed_file filename`  
The bot will read file from its current folder, parse it as embed json code, then send embed message to the channel.

You can make your embed code at [Leovoel's Embed Visualizer](https://leovoel.github.io/embed-visualizer/).

P.S. If `author` field in embed code is not given, the bot will use the one who send `!embed` command's avatar and display name with author.

CHANGELOG.md
# 2018 / 11 / 23 1.1.3 Fix Deimos Bug
* Fix deimos's health can't be calculate bug.

# 2018 / 11 / 23 1.1.2 Add -healthover
* Add `-healthover` command.
* Known issue: can't determine Deimos's encounter result.

# 2018 / 10 / 31 1.1.1 Bug Fix

# 2018 / 10 / 25 1.1.0 Features Update
* Add commands `-raidarwaittime`, `-raidarretrycount`, `-raidarsearchcount`, `-notonlyraidar`
* Remove `Gw2RaidarAcceptable` filed from `BossList.json`
* Add optional field `LifeThreshold` into `BossList.json`
* Get Gw2Raidar support list before parsing, now the tool will base on this support decide whether upload the log or not.

EmbedHelper.py
import discord  #python -m pip install -U git+https://github.com/Rapptz/discord.py@rewrite
from discord.ext import commands
import json
import sys

bot = commands.Bot(command_prefix="!", description="Discord Embed Helper")

@bot.event
async def on_ready():
  print('Logged in as')
  print(bot.user.name)
  print(bot.user.id)
  print('------')

@bot.command()
async def embed(ctx):
  try:
    embed = json.loads(ctx.message.content[7:])
    if "author" not in embed:
      embed["author"] = dict()
      embed["author"]["name"] = ctx.message.author.display_name
      embed["author"]["icon_url"] = ctx.message.author.avatar_url        
    await ctx.message.delete()        
    await ctx.send(embed = discord.Embed.from_data(embed))
  except BaseException as e:
    await ctx.send(str(e))

@bot.command()
async def embed_file(ctx, filename):
  try:
    with open(filename, "r") as inputFile:
      embed = json.load(inputFile)
    if "author" not in embed:
      embed["author"] = dict()
      embed["author"]["name"] = ctx.message.author.display_name
      embed["author"]["icon_url"] = ctx.message.author.avatar_url        
    await ctx.message.delete()        
    await ctx.send(embed = discord.Embed.from_data(embed))
  except BaseException as e:
    await ctx.send(str(e))


with open("Config.json") as configFile:
  config = json.load(configFile)
if "DiscordBotToken" not in config:
  print("Can't find \"DiscordBotToken\" in Config.json!")

bot.run(config["DiscordBotToken"])

EvtcParser.py
from collections import OrderedDict
import zipfile
import json
import sys
import os

class Iff:
  IFF_FRIEND   = 0
  IFF_FOE      = 1
  IFF_UNKNOWN  = 2 # or uncertain

class CbtResult:
  CBTR_NORMAL      = 0 # good physical hit
  CBTR_CRIT        = 1 # physical hit was crit
  CBTR_GLANCE      = 2 # physical hit was glance
  CBTR_BLOCK       = 3 # physical hit was blocked eg. mesmer shield 4
  CBTR_EVADE       = 4 # physical hit was evaded, eg. dodge or mesmer sword 2
  CBTR_INTERRUPT   = 5 # physical hit interrupted something
  CBTR_ABSORB      = 6 # physical hit was "invlun" or absorbed eg. guardian elite
  CBTR_BLIND       = 7 # physical hit missed
  CBTR_KILLINGBLOW = 8 # hit was killing hit
  CBTR_DOWNED      = 9 # hit was downing hit
 
class CbtActivation:
  ACTV_NONE          = 0 # not used - not this kind of event
  ACTV_NORMAL        = 1 # activation without quickness
  ACTV_QUICKNESS     = 2 # activation with quickness
  ACTV_CANCEL_FIRE   = 3 # cancel with reaching channel time
  ACTV_CANCEL_CANCEL = 4 # cancel without reaching channel time
  ACTV_RESET         = 5 # animation completed fully

class CbtStateChange:
  CBTS_NONE            = 0  # not used - not this kind of event
  CBTS_ENTERCOMBAT     = 1  # src_agent entered combat, dst_agent is subgroup
  CBTS_EXITCOMBAT      = 2  # src_agent left combat
  CBTS_CHANGEUP        = 3  # src_agent is now alive
  CBTS_CHANGEDEAD      = 4  # src_agent is now dead
  CBTS_CHANGEDOWN      = 5  # src_agent is now downed
  CBTS_SPAWN           = 6  # src_agent is now in game tracking range
  CBTS_DESPAWN         = 7  # src_agent is no longer being tracked
  CBTS_HEALTHUPDATE    = 8  # src_agent has reached a health marker. dst_agent = percent * 10000 (eg. 99.5% will be 9950)
  CBTS_LOGSTART        = 9  # log start. value = server unix timestamp **uint32**. buff_dmg = local unix timestamp. src_agent = 0x637261 (arcdps id)
  CBTS_LOGEND          = 10 # log end. value = server unix timestamp **uint32**. buff_dmg = local unix timestamp. src_agent = 0x637261 (arcdps id)
  CBTS_WEAPSWAP        = 11 # src_agent swapped weapon set. dst_agent = current set id (0/1 water, 4/5 land)
  CBTS_MAXHEALTHUPDATE = 12 # src_agent has had it's maximum health changed. dst_agent = new max health
  CBTS_POINTOFVIEW     = 13 # src_agent will be agent of "recording" player
  CBTS_LANGUAGE        = 14 # src_agent will be text language
  CBTS_GWBUILD         = 15 # src_agent will be game build
  CBTS_SHARDID         = 16 # src_agent will be sever shard id
  CBTS_REWARD          = 17 # src_agent is self, dst_agent is reward id, value is reward type. these are the wiggly boxes that you get
  CBTS_BUFFINITIAL     = 18 # combat event that will appear once per buff per agent on logging start (zero duration, buff==18)
  CBTS_POSITION        = 19 # src_agent changed, cast float* p = (float*)&dst_agent, access as x/y/z (float[3])
  CBTS_VELOCITY        = 20 # src_agent changed, cast float* v = (float*)&dst_agent, access as x/y/z (float[3])
  CBTS_FACING          = 21 # src_agent changed, cast float* f = (float*)&dst_agent, access as x/y (float[2])
  CBTS_TEAMCHANGE      = 22 # src_agent change, dst_agent new team id  

class CbtBuffRemove:
  CBTB_NONE   = 0 # not used - not this kind of event
  CBTB_ALL    = 2 # last/all stacks removed (sent by server)
  CBTB_SINGLE = 3 # single stack removed (sent by server). will happen for each stack on cleanse
  CBTB_MANUAL = 4 # single stack removed (auto by arc on ooc or all stack, ignore for strip/cleanse calc, use for in/out volume)
 
class CbtCustomSkill:
  CSK_RESURRECT = 1066, # not custom but important and unnamed
  CSK_BANDAGE   = 1175, # personal healing only
  CSK_DODGE     = 65001 # will occur in is_activation==normal event

class GwLanguage:
  GWL_ENG = 0,
  GWL_FRE = 2,
  GWL_GEM = 3,
  GWL_SPA = 4,

class Agent:
  LEN = 96
  def __init__(self, bytes):
    self.addr          = int.from_bytes(bytes[0:8], byteorder="little", signed=False)
    self.porf          = int.from_bytes(bytes[8:12], byteorder="little", signed=False)
    self.isElite       = int.from_bytes(bytes[12:16], byteorder="little", signed=False)
    self.toughness     = int.from_bytes(bytes[16:18], byteorder="little", signed=False)
    self.concentration = int.from_bytes(bytes[18:20], byteorder="little", signed=False)
    self.healing       = int.from_bytes(bytes[20:22], byteorder="little", signed=False)
    self.pad1          = int.from_bytes(bytes[22:24], byteorder="little", signed=False)
    self.condition     = int.from_bytes(bytes[24:26], byteorder="little", signed=False)
    self.pad2          = int.from_bytes(bytes[26:28], byteorder="little", signed=False)
    names = bytes[28:92].decode("utf8").split("\x00")
    self.name = names[0]
    if self.isPlayer:
      self.displayName = names[1][1:]
      self.subgroup = int(names[2])


  @property
  def isgadget(self):
    if self.isElite != 0xFFFFFFFF:
      return False 
    if self.porf >> 8 != 0xFFFF:
      return False
    return True
  
  @property
  def isNpc(self):
    if self.isElite != 0xFFFFFFFF:
      return False 
    if self.porf >> 8 == 0xFFFF:
      return False
    return True

  @property
  def isPlayer(self):
    if self.isElite == 0xFFFFFFFF:
      return False 
    return True
  
  @property
  def specialID(self):
    return self.porf & 0xFFFF


class Skill:
  LEN = 68
  def __init__(self, bytes):
    self.id = int.from_bytes(bytes[0:4], byteorder="little", signed=False)
    self.name = bytes[4:68].decode("utf8").split("\x00")[0]


class CombatEvent0:
  LEN = 8 * 3 + 4 * 2 + 2 * 5 + 22
  def __init__(self, bytes):
    self.time                  = int.from_bytes(bytes[0:8], byteorder="little", signed=False)    #8 byte
    self.src_agent             = int.from_bytes(bytes[8:16], byteorder="little", signed=False)   #8 byte
    self.dst_agent             = int.from_bytes(bytes[16:24], byteorder="little", signed=False)  #8 byte
    self.value                 = int.from_bytes(bytes[24:28], byteorder="little", signed=False)  #4 byte
    self.buff_dmg              = int.from_bytes(bytes[28:32], byteorder="little", signed=False)  #4 byte
    self.overstack_value       = int.from_bytes(bytes[32:34], byteorder="little", signed=False)  #2 byte
    self.skillid               = int.from_bytes(bytes[34:36], byteorder="little", signed=False)  #2 byte
    self.src_instid            = int.from_bytes(bytes[36:38], byteorder="little", signed=False)  #2 byte
    self.dst_instid            = int.from_bytes(bytes[38:40], byteorder="little", signed=False)  #2 byte
    self.src_master_instid     = int.from_bytes(bytes[40:42], byteorder="little", signed=False)  #2 byte
    self.iss_offset            = int(bytes[42])
    self.iss_offset_target     = int(bytes[43])
    self.iss_bd_offset         = int(bytes[44])
    self.iss_bd_offset_target  = int(bytes[45])
    self.iss_alt_offset        = int(bytes[46])
    self.iss_alt_offset_target = int(bytes[47])
    self.skar                  = int(bytes[48])
    self.skar_alt              = int(bytes[49])
    self.skar_use_alt          = int(bytes[50])
    self.iff                   = int(bytes[51])
    self.buff                  = int(bytes[52])
    self.result                = int(bytes[53])
    self.is_activation         = int(bytes[54])
    self.is_buffremove         = int(bytes[55])
    self.is_ninety             = int(bytes[56])
    self.is_fifty              = int(bytes[57])
    self.is_moving             = int(bytes[58])
    self.is_statechange        = int(bytes[59])
    self.is_flanking           = int(bytes[60])
    self.is_shields            = int(bytes[61])
    self.is_offcycle           = int(bytes[62])
    self.pad64                 = int(bytes[63])

class CombatEvent1:
  LEN = 8 * 3 + 4 * 4 + 2 * 4 + 16
  def __init__(self, bytes):
    self.time = int.from_bytes(bytes[0:8], byteorder="little", signed=False)                 #8 byte
    self.src_agent         = int.from_bytes(bytes[8:16], byteorder="little", signed=False)   #8 byte
    self.dst_agent         = int.from_bytes(bytes[16:24], byteorder="little", signed=False)  #8 byte
    self.value             = int.from_bytes(bytes[24:28], byteorder="little", signed=True)   #4 byte
    self.buff_dmg          = int.from_bytes(bytes[28:32], byteorder="little", signed=True)   #4 byte
    self.overstack_value   = int.from_bytes(bytes[32:36], byteorder="little", signed=False)  #4 byte
    self.skillid           = int.from_bytes(bytes[36:40], byteorder="little", signed=False)  #4 byte
    self.src_instid        = int.from_bytes(bytes[40:42], byteorder="little", signed=False)  #2 byte
    self.dst_instid        = int.from_bytes(bytes[42:44], byteorder="little", signed=False)  #2 byte
    self.src_master_instid = int.from_bytes(bytes[44:46], byteorder="little", signed=False)  #2 byte
    self.dst_master_instid = int.from_bytes(bytes[46:48], byteorder="little", signed=False)  #2 byte
    self.iff               = int(bytes[48])
    self.buff              = int(bytes[49])
    self.result            = int(bytes[50])
    self.is_activation     = int(bytes[51])
    self.is_buffremove     = int(bytes[52])
    self.is_ninety         = int(bytes[53])
    self.is_fifty          = int(bytes[54])
    self.is_moving         = int(bytes[55])
    self.is_statechange    = int(bytes[56])
    self.is_flanking       = int(bytes[57])
    self.is_shields        = int(bytes[58])
    self.is_offcycle       = int(bytes[59])
    self.pad61             = int(bytes[60])
    self.pad62             = int(bytes[61])
    self.pad63             = int(bytes[62])
    self.pad63             = int(bytes[63])


class EvtcLog:
  def __init__(self, filepath, quickParse=False, lifeThreshold = 50):
    if zipfile.is_zipfile(filepath):
      with zipfile.ZipFile(filepath) as zipFile:
        namelist = zipFile.namelist()
        for n in namelist:
          if n.endswith(".evtc") or n.endswith(".evtc.tmp"):
            filename = n
        try:
          filename
        except:
          raise BaseException("can't find evtc file in zip file {}".format(filepath))
        with zipFile.open(filename) as evtcFile:
          self.parseEvtc(evtcFile, lifeThreshold, zipFile.infolist()[0].file_size, quickParse)
    else:
      with open(filepath, "rb") as evtcFile:
        self.parseEvtc(evtcFile, lifeThreshold, os.path.getsize(filepath), quickParse)
  
  def parseEvtc(self, evtc, lifeThreshold, size, quickParse):
    if evtc.read(4).decode("ascii") != "EVTC":
      raise BaseException("Input file isn't evtc file or standard zip.")
    self.dateText = evtc.read(8).decode("ascii")
    self.revision = evtc.read(1)[0]
    self.bossId = int.from_bytes(evtc.read(2), byteorder="little", signed=False)
    #evtc.seek(16)
    evtc.read(1)
    bossIdForResult = self.bossId

    self.cbtResult = False
    self.playerNames = []
    self.agentCount = int.from_bytes(evtc.read(4), byteorder="little", signed=False)
    self.agents = []
    self.finalHealth = 10000
    self.bossName = ""
    playersAddr = []
    bossAddr = None
    for i in range(0, self.agentCount):
      agent = Agent(evtc.read(Agent.LEN))
      self.agents.append(agent)
      if agent.isPlayer:
        playersAddr.append(agent.addr)
        self.playerNames.append(agent.displayName)
        self.playerNames.append(agent.name)
      else:
        if bossAddr == None and agent.specialID == bossIdForResult:
          bossAddr = agent.addr
        if bossAddr == None and agent.specialID == self.bossId:
          self.bossName = agent.name

    self.skillCount = int.from_bytes(evtc.read(4), byteorder="little", signed=False)
    self.skills = []
    for i in range(0, self.skillCount):
      self.skills.append(Skill(evtc.read(Skill.LEN)))
    if self.revision == 0:
      CombatEvent = CombatEvent0
    else:
      CombatEvent = CombatEvent1
    if not quickParse:
      self.combatEvents = []
      data = evtc.read(CombatEvent.LEN)
      while len(data) == CombatEvent.LEN:
        evtcLog = CombatEvent(data)
        isBoss = False
        for agent in self.agents:
          if agent.addr == evtcLog.src_agent:
            if self.bossId == 16246:   #For Xera
                if agent.isNpc and (agent.specialID == 16286 or agent.specialID == self.bossId):
                  isBoss = True
            elif self.bossId == 17154: #For Deimos
                if agent.isNpc and (agent.name.startswith("at24660") or agent.specialID == self.bossId):
                  isBoss = True
            else:
              if agent.specialID == self.bossId:
                isBoss = True
            break

        if isBoss:
          if not self.cbtResult:
            if evtcLog.is_statechange == CbtStateChange.CBTS_CHANGEDEAD:
              self.cbtResult = True
            elif lifeThreshold >= 0 and evtcLog.is_statechange == CbtStateChange.CBTS_HEALTHUPDATE and evtcLog.dst_agent < lifeThreshold:
              self.cbtResult = True
          if evtcLog.is_statechange == CbtStateChange.CBTS_HEALTHUPDATE:
            self.finalHealth = evtcLog.dst_agent
        self.combatEvents.append(evtcLog)
        data = evtc.read(CombatEvent.LEN)

      self.combatTimeUsed = self.combatEvents[-1].time - self.combatEvents[0].time
    #seek is broken in python 3.7
    #else:
    #  data = evtc.read(CombatEvent.LEN)
    #  firstLog = CombatEvent(data)
    #  evtc.seek(size - CombatEvent.LEN)
    #  data = evtc.read(CombatEvent.LEN)
    #  lastLog = CombatEvent(data)
    #  self.combatTimeUsed = lastLog.time - firstLog.time

if __name__ == "__main__":
  if len(sys.argv) < 2:
    print(sys.argv[0], "EvtcFile [-json]: dump evtc file.")
    print(sys.argv[0], "SuccessEvtcFile -config: config specific boss in BossList by given log.")
    sys.exit(0)
  log = EvtcLog(sys.argv[1])
  lastLifeChange = log.finalHealth
  bossAddr = 0
  isDead = False
  for agent in log.agents:
    if not agent.isPlayer:
      if agent.specialID == log.bossId:
        bossAddr = agent.addr
        break
  for event in log.combatEvents:
    if event.is_statechange == CbtStateChange.CBTS_CHANGEDEAD and event.src_agent == bossAddr:
      isDead = True
  if len(sys.argv) > 2 and sys.argv[2] == "-config":
    with open("BossList.json", "r") as listFile:
      bossList = json.load(listFile)
    for boss in bossList["Bosses"]:
      if boss["Name"] == log.bossName:
        if isDead:
          if "LifeThreshold" in boss:
            del boss["LifeThreshold"]
        else:
          boss["LifeThreshold"] = lastLifeChange + 30
        with open("BossList.json", "w") as listFile:
          json.dump(bossList, listFile, indent=2)
        sys.exit(0)
  else:
    if len(sys.argv) > 2 and sys.argv[2] == "-json":
      fileName = "EvtcLog.json"
    else:
      fileName = "EvtcLog.txt"
    with open(fileName, "w", encoding="utf8") as output:
      if len(sys.argv) > 2 and sys.argv[2] == "-json":
        outputDict = OrderedDict()
        outputDict["BossId"] = log.bossId
        outputDict["LastLifeChange"] = lastLifeChange
        outputDict["IsBossDead"] = isDead
        outputDict["Agents"] = []
        for agents in log.agents:
          outputDict["Agents"].append(agents.__dict__)
        json.dump(outputDict, output, indent=2)
        outputDict["Skills"] = []
        for skill in log.skills:
          outputDict["Skills"].append(skill.__dict__)
        outputDict["Events"] = []
        for event in log.combatEvents:
          outputDict["Events"].append(event.__dict__)
        json.dump(outputDict, output, indent=2)
      else:
        print("BossId:", log.bossId, file=output)
        print("Last Life Change:", lastLifeChange, file=output)
        print("Is Boss Dead:", isDead, file=output)
        print("Agents:", file=output)
        for agent in log.agents:
          print(" ", agent.name, end=" ", file=output)
          if agent.isPlayer:
            print(agent.displayName, agent.addr, file=output)
          else:
            print(agent.specialID, agent.addr, file=output)            
        print("Events:", file=output)
        for event in log.combatEvents:
          srcName = ""
          dstName = ""
          for agent in log.agents:
            if agent.addr == event.src_agent:
              srcName = agent.name
            if agent.addr == event.dst_agent:
              dstName = agent.name
          print("  Src:", srcName, event.src_agent, "Dst:", dstName, event.dst_agent, "StateChange:", event.is_statechange, file=output)
      
